cypress-io/cypress
# fix: Improve detection of programmatic scrolls to avoid unintended disabling of auto scrolling

- Closes #25084 

### Additional details
(First time contributor, please be gentle)

Fixes an issue where multiple "programmatic" scrolls in quick succession causes auto scroll to be disabled when it shouldn't have been.

The previous Scroller implementation used a simple boolean toggle (`_userScroll: boolean`) for signalling that the _next_ `scroll` event had been programmatically dispatched:

1. `scrollIntoView()` sets `_userScroll = true` and fires a `scroll` event
2. the `scroll` handler recognises this and takes no further action other than resetting `_userScroll = false`

This approach would fail if multiple calls to `scrollIntoView()` happen before the corresponding `scroll` events could be queued and handled by the browser.  Only the first event would be treated as a programmatic scroll, and all others would be incorrectly treated as user scrolls.

The new implementation eschews the `_userScroll` boolean entirely, and instead uses the existing `_userScrollCount` property to track programmatic vs user scrolls:

1. `scrollIntoView()` decrements `_userScrollCount` by 1 each time it is called
2. the `scroll` handler increments `_userScrollCount` by 1 each time an event is handled

This means that the _order_ in which programmatic scroll events are created and handled is no longer critical.

Instead, it is the _number_ of `scroll` events minus the count of `scrollIntoView()` calls that determines when to disable auto scroll.

I chose to retain the existing _"3 or more user scrolls within 50ms"_ logic from the previous implementation.

### Steps to test
The existing unit tests for the scroller have been expanded to include a scenario where 2 x `scrollIntoView()` calls are made, followed by 4 x 'scroll' events being handled (over a period > 50ms).  Previously this would have disabled auto scrolling, but as `4 - 2 = 2`, which is under the "3 or more user scrolls are needed" threshold, auto scrolling will no longer disable.

Part of the reason #25084 has remained open for so long was the inability to reliably reproduce the issue. I have a project that can reliably reproduce the problem, and I tested this change against that project and can confirm that it resolves the issue.

### How has the user experience changed?
No change, other than auto scrolling no longer being disabled without the user's intent.

### PR Tasks
